### PR TITLE
docs: add transmitted  blog of APISIX&Nginx

### PR DESCRIPTION
**Pull request by SylviaBABY**
Changes:

add a transmitted tech blog of apisix&Nginx

### CODE REVIEW

File: website/blog/2021-08-10-apisix-nginx.md [hunk 1]
```diff
@@ -0,0 +1,512 @@
+---
+title: "Apache APISIX 架构分析：如何动态管理 Nginx 集群？"
+keywords: 
+- APISIX
+- Nginx
+- 动态管理
+- Lua
```
Comment (yzeng25) [inline website/blog/2021-08-10-apisix-nginx.md]:
```
- API 网关
- Apache APISIX
- Nginx
- Lua
- 动态管理
```

File: website/blog/2021-08-10-apisix-nginx.md [hunk 2]
```diff
@@ -0,0 +1,512 @@
...
+description: 本文转发自陶辉个人博客，主要介绍了基于APISIX 2.8 版本、OpenResty 1.19.3.2 版本以及 Nginx 1.19.3 版本进行 Apache APISIX 实现 REST API 远程控制 Nginx 集群的原理讲解。
+tags: [technology]
```
Comment (yzeng25) [inline website/blog/2021-08-10-apisix-nginx.md:11]: add date

File: website/blog/2021-08-10-apisix-nginx.md [hunk 3]
```diff
@@ -0,0 +1,512 @@
...
+* 多进程架构增大了 Worker 进程间的数据同步难度，必须选择 1 个低成本的实现机制，保证每个 Nginx 节点、Worker 进程都持有最新的配置
+
+Apache APISIX 基于 Lua 定时器及 lua-resty-etcd 模块实现了配置的动态管理，本文将基于APISIX 2.8 版本、OpenResty 1.19.3.2 版本以及 Nginx 1.19.3 版本进行 Apache APISIX 实现 REST API 远程控制 Nginx 集群的原理。
+
+## 基于 ETCD watch 机制的配置同步方案
+
+管理集群必须依赖中心化的配置，etcd 就是这样一个数据库。Apache APISIX 没有选择关系型数据库作为配置中心，是因为 etcd 具有以下 2 个优点：
+
+* etcd 采用类 Paxos 的 Raft 协议保障了数据一致性，它是去中心化的分布式数据库，可靠性高于关系数据库
+* etcd 的 watch 机制允许客户端监控某个 key 的变动，即，若类似 /nginx/http/upstream 这种key 的 value 值发生变动，watch 客户端会立刻收到通知，如下图所示：
```
Comment (yzeng25) [inline website/blog/2021-08-10-apisix-nginx.md]: 这种 key

File: website/blog/2021-08-10-apisix-nginx.md [hunk 4]
```diff
@@ -0,0 +1,512 @@
...
+![基于 etcd 同步 nginx 配置](https://static.apiseven.com/202108/1631170345853-f020a64d-3e97-49c0-8395-c9e4e9cf4233.jpeg)
+
+因此，不同于 [Orange](https://github.com/orlabs/orange) 采用 MySQL、[Kong](https://konghq.com/) 采用 PostgreSQL 作为配置中心的方式（这二者同样是基于 OpenResty 实现的 API Gateway），Apache APISIX 采用了 etcd 作为中心化的配置组件。
```
Comment (yzeng25) [inline website/blog/2021-08-10-apisix-nginx.md]: “不同于 orange 和 kong”

File: website/blog/2021-08-10-apisix-nginx.md [hunk 5]
```diff
@@ -0,0 +1,512 @@
...
+```yaml
+etcd:
+  host:  
+    - "http://127.0.0.1:2379"
+  prefix: /apisix                 # apisix configurations prefix
+```
+
+而 upstreams/1 就等价于 nginx.conf 中的 http { upstream 1 {} } 配置。类似关键字还有 /apisix/services/、/apisix/routes/ 等。
+
+那么，Nginx 是怎样通过 watch 机制获取到 etcd 配置数据变化的呢？有没有新启动一个 agent进程？它通过 HTTP/1.1 还是 gRPC 与 etcd 通讯的？
```
Comment (yzeng25) [inline website/blog/2021-08-10-apisix-nginx.md]: agent 进程

File: website/blog/2021-08-10-apisix-nginx.md [hunk 6]
```diff
@@ -0,0 +1,512 @@
...
+
+## ngx.timer.at 定时器
+
+Apache APISIX 并没有启动 Nginx 以外的进程与 etcd 通讯。实际上它是通过 `ngx.timer.at` 这个定时器实现了 watch 机制。为了方便对 OpenResty 不太了解的同学，我们先来看看 Nginx 中的定时器是如何实现的，它是 watch 机制实现的基础。
```
Comment (yzeng25) [inline website/blog/2021-08-10-apisix-nginx.md]: 为了方便对 OpenResty 不太了解的同学理解

File: website/blog/2021-08-10-apisix-nginx.md [hunk 7]
```diff
@@ -0,0 +1,512 @@
...
+    local ok, err = xpcall(function()
+        local ok, err = sync_data(self)
+    end, debug.traceback)
+    ngx_timer_at(0, _automatic_fetch, self)
+end
+```
+
+`sync_data` 函数将通过 etcd 的 watch 机制获取更新，它的实现机制我们接下来会详细分析。
+
+所以总结来看，APISIX 在每个 Nginx Worker 进程的启动过程中，通过 `ngx.timer.at` 函数将 `_automatic_fetch` 插入定时器。`_automatic_fetch` 函数执行时会通过 `sync_data` 函数，基于 watch 机制接收 etcd 中的配置变更通知，这样，每个 Nginx 节点、Worker 进程都将保持最新的配置。如此设计还有 1 个明显的优点：etcd 中的配置直接写入 Nginx Worker 进程中，这样处理请求时就能直接使用新配置，无须在进程间同步配置，这要比启动 1 个 agent 进程更简单！
```
Comment (yzeng25) [inline website/blog/2021-08-10-apisix-nginx.md]: Apache APISIX

File: website/blog/2021-08-10-apisix-nginx.md [hunk 8]
```diff
@@ -0,0 +1,512 @@
...
+        ssl_verify = self.ssl_verify,
+        ssl_cert_path = self.ssl_cert_path,
+        ssl_key_path = self.ssl_key_path,
+    })
+
+    return endpoint, err
+end
+```
+
+可见，APISIX 在每个 worker 进程中，**通过 `ngx.timer.at` 和 lua-resty-etcd 库反复请求 etcd，以此保证每个 Worker 进程中都含有最新的配置。**
```
Comment (yzeng25) [inline website/blog/2021-08-10-apisix-nginx.md]: Apache APISIX

File: website/blog/2021-08-10-apisix-nginx.md [hunk 9]
```diff
@@ -0,0 +1,512 @@
...
+
+## APISIX 配置与插件的远程变更
```
Comment (yzeng25) [inline website/blog/2021-08-10-apisix-nginx.md:316]: ditto

File: website/blog/2021-08-10-apisix-nginx.md [hunk 10]
```diff
@@ -0,0 +1,512 @@
...
+
+我们当然可以直接通过 gRPC 接口修改 etcd 中相应 key 的内容，再基于上述的 watch 机制使得 Nginx 集群自动更新配置。然而，这样做的风险很大，因为配置请求没有经过校验，进面导致配置数据与 Nginx 集群不匹配。
+
+### 通过 Nginx 的 /apisix/admin/ 接口修改配置
+
+APISIX 提供了这么一种机制：访问任意 1 个 Nginx 节点，通过其 Worker 进程中的 Lua 代码校验请求成功后，再由 /v3/dv/put 接口写入 etcd 中。下面我们来看看 APISIX 是怎么实现的。
+
+首先，make run 生成的 nginx.conf 会自动监听 9080 端口（可通过 config.yaml 中 apisix.node_listen 配置修改），当 `apisix.enable_admin` 设置为 true 时，nginx.conf 就会生成以下配置：
+
+```conf
```
Comment (yzeng25) [inline website/blog/2021-08-10-apisix-nginx.md]: "```yaml"

File: website/blog/2021-08-10-apisix-nginx.md [hunk 11]
```diff
@@ -0,0 +1,512 @@
...
+        methods = {"PUT"},
+        handler = post_reload_plugins,
+    },
+}
+```
+
+比如，当通过 /apisix/admin/upstreams/1 和 PUT 方法创建 1 个 Upstream 上游时：
+
+```shell
+# curl "http://127.0.0.1:9080/apisix/admin/upstreams/1" -H "X-API-KEY: edd1c9f034335f136f87ad84b625c8f1" -X PUT -d '
```
Comment (yzeng25) [inline website/blog/2021-08-10-apisix-nginx.md]: delete ”# “

File: website/blog/2021-08-10-apisix-nginx.md [hunk 12]
```diff
@@ -0,0 +1,514 @@
...
+
+![APISIX 架构图](https://static.apiseven.com/202108/1631170283612-ba5e27ff-726b-47a6-aa51-84731b067c44.png)
+
+让 Nginx 具备动态、集群管理能力并不容易，因为这将面临以下问题：
+
+* 微服务架构使得上游服务种类多、数量大，这导致路由规则、上游 Server 的变更极为频率。而 Nginx 的路由匹配是基于静态的 Trie 前缀树、哈希表、正则数组实现的，一旦`server_name`、`location` 变动，不执行 reload 就无法实现配置的动态变更
+* Nginx 将自己定位于 ADC 边缘负载均衡，因此它对上游并不支持 HTTP2 协议。这增大了 OpenResty 生态实现 etcd gRPC 接口的难度，因此通过 watch 机制接收配置变更必然效率低下
+* 多进程架构增大了 Worker 进程间的数据同步难度，必须选择 1 个低成本的实现机制，保证每个 Nginx 节点、Worker 进程都持有最新的配置
+
+Apache APISIX 基于 Lua 定时器及 lua-resty-etcd 模块实现了配置的动态管理，本文将基于APISIX 2.8 版本、OpenResty 1.19.3.2 版本以及 Nginx 1.19.3 版本进行 Apache APISIX 实现 REST API 远程控制 Nginx 集群的原理。
```
Comment (Serendipity96) [inline website/blog/2021-08-10-apisix-nginx.md]: add a space, 本文将基于APISIX 2.8 版本 -> 本文将基于 APISIX 2.8 版本

File: website/blog/2021-08-10-apisix-nginx.md [hunk 13]
```diff
@@ -0,0 +1,514 @@
...
+
+## 基于 ETCD watch 机制的配置同步方案
```
Comment (Serendipity96) [inline website/blog/2021-08-10-apisix-nginx.md]: Would you like to uniform capitalization "etcd" in article？

File: website/blog/2021-08-10-apisix-nginx.md [hunk 14]
```diff
@@ -0,0 +1,514 @@
...
+    ev->handler = ngx_http_lua_timer_handler;
+    ngx_add_timer(ev, delay);
+}
+```
+
+因此，当我们调用 `ngx.timer.at`  Lua 定时器时，就是在 Nginx 的红黑树定时器里加入了 `ngx_http_lua_timer_handler` 回调函数，这个函数不会阻塞 Nginx。
+
+下面我们来看看 Apache APISIX 是怎样使用 `ngx.timer.at` 的。
+
+### APISIX 基于定时器实现的 watch 机制
```
Comment (Serendipity96) [inline website/blog/2021-08-10-apisix-nginx.md]: Apache APISIX

File: website/blog/2021-08-10-apisix-nginx.md [hunk 15]
```diff
@@ -0,0 +1,514 @@
...
+
+    local_conf_path = profile:yaml_path("config")
+    local user_conf_yaml, err = util.read_file(local_conf_path)
+    ok, err = merge_conf(default_conf, user_conf)
+end
+```
+
+可见，ngx_tpl.lua 模板中仅部分数据可由 yaml 配置中替换，其中 conf/config-default.yaml 是官方提供的默认配置，而 conf/config.yaml 则是由用户自行覆盖的自定义配置。如果你觉得仅替换模板数据还不够，大可直接修改 ngx_tpl 模板。
+
+#### APISIX 获取 etcd 通知的方式
```
Comment (Serendipity96) [inline website/blog/2021-08-10-apisix-nginx.md]: Apache APISIX

File: website/blog/2021-08-10-apisix-nginx.md [hunk 16]
```diff
@@ -0,0 +1,514 @@
...
+
+最后再提下 Script，它与 Plugin 是互斥的。之前的动态调整改的只是配置，事实上 Lua JIT 的及时编译还提供了另外一个杀手锏 loadstring，它可以将字符串转换为 Lua 代码。因此，在 etcd 中存储 Lua 代码并设置为 Script 后，就可以将其传送到 Nginx 上处理请求了。
+
+## 总结
+
+Nginx 集群的管理必须依赖中心化配置组件，而高可靠又具备 watch 推送机制的 etcd 无疑是最合适的选择！虽然当下 Resty 生态没有 gRPC 客户端，但每个 Worker 进程直接通过 HTTP/1.1 协议同步 etcd 配置仍不失为一个好的方案。
+
+动态修改 Nginx 配置的关键在于2点：Lua 语言的灵活度远高于 nginx.conf 语法，而且 Lua 代码可以通过 loadstring 从外部数据中导入。当然，为了保障路由匹配的执行效率，Apache APISIX 通过 C 语言实现了前缀基数树，基于 Host、Method、URI 进行请求匹配，在保障动态性的基础上提升了性能。
+
+Apache APISIX 拥有许多优秀的设计，本文仅讨论了 Nginx 集群的动态管理。
```
Comment (yzeng25) [inline website/blog/2021-08-10-apisix-nginx.md:512]: could you add the origin post link from Tao Hui's blog?

Comment (SylviaBABY) [reply website/blog/2021-08-10-apisix-nginx.md:512]: done

**Review by yzeng25 — APPROVED** [review]

File: website/blog/2021-08-10-apisix-nginx.md [hunk 17]
```diff
@@ -0,0 +1,516 @@
+---
+title: "Apache APISIX 架构分析：如何动态管理 Nginx 集群？"
+keywords: 
+- API 网关
+- Apache APISIX
+- Nginx
+- - Lua
```
Comment (bzp2010) [inline website/blog/2021-08-10-apisix-nginx.md]: Please correct the format.

**Review by bzp2010 — CHANGES_REQUESTED** [review]

File: website/blog/2021-08-10-apisix-nginx.md (hunk 17 above)
Comment (SylviaBABY) [reply website/blog/2021-08-10-apisix-nginx.md]: done

**Review by bzp2010 — APPROVED** [review]

File: website/blog/2021-08-10-apisix-nginx.md [hunk 18]
```diff
@@ -0,0 +1,516 @@
...
+
+管理集群必须依赖中心化的配置，etcd 就是这样一个数据库。Apache APISIX 没有选择关系型数据库作为配置中心，是因为 etcd 具有以下 2 个优点：
+
+* etcd 采用类 Paxos 的 Raft 协议保障了数据一致性，它是去中心化的分布式数据库，可靠性高于关系数据库
+* etcd 的 watch 机制允许客户端监控某个 key 的变动，即，若类似 /nginx/http/upstream 这种 key 的 value 值发生变动，watch 客户端会立刻收到通知，如下图所示：
+![基于 etcd 同步 nginx 配置](https://static.apiseven.com/202108/1631170345853-f020a64d-3e97-49c0-8395-c9e4e9cf4233.jpeg)
+
+因此，不同于 Orange 和 Kong，Apache APISIX 采用了 etcd 作为中心化的配置组件。
+
+因此，你可以在生产环境的 Apache APISIX 中通过 etcdctl 看到如下类似配置：
```
Comment (liuxiran) [inline website/blog/2021-08-10-apisix-nginx.md]: 两个 `因此`，去掉一个？

**Review by liuxiran — APPROVED** [review]

File: website/blog/2021-08-10-apisix-nginx.md (hunk 18 above)
Comment (SylviaBABY) [reply website/blog/2021-08-10-apisix-nginx.md]: done

**Pull request merged**